### PR TITLE
Add ActuatorHandle and Implement string-based interface handle-handling using DynamicJointState message

### LINKED ISSUE #85: Implement registerInterface() following a string-based setup so it can be checked against when attempting to load transmissions
https://github.com/ros-controls/roadmap/blob/master/design_drafts/joint_interfaces_and_transmissions.md

### PR DESCRIPTION
**Pull request by bmagyar**
This PR adds actuator handles and interfaces for handling them. 
It is all by using strings so it should support pretty much anything. The backing data structure is a DynamicJointState message which seems to be quite useful for this although a map-map structure may fare better, but I consider that an implementation improvement, the current code implements the functionality required by a transmission parser / loader which is coming soon.

There's extensive testing for all the basic cases.

What this PR *doesn't have* is refactoring all the functions using joint handles. If this approach is favoured, we could do that in a separate PR.

Since the role of handles are to remain pretty much the same this shouldn't conceptually conflict with #101 .

resolves #85

### CODE REVIEW

File: hardware_interface/test/test_register_actuators.cpp [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+  EXPECT_EQ(actuator_handle.get_interface_name(), FOO_INTERFACE);
+
+  EXPECT_EQ(HW_RET_OK, robot_hw.register_actuator(ACTUATOR_NAME, BAR_INTERFACE));
+  EXPECT_THAT(robot_hw.get_registered_actuator_names(), ElementsAre(ACTUATOR_NAME));
+
+  EXPECT_EQ(HW_RET_OK, robot_hw.register_actuator(ACTUATOR2_NAME, FOO_INTERFACE));
+  EXPECT_THAT(
+    robot_hw.get_registered_actuator_names(),
+    UnorderedElementsAre(ACTUATOR_NAME, ACTUATOR2_NAME));
+}
```
Comment (jordan-palacios) [inline hardware_interface/test/test_register_actuators.cpp:125]: There is no test that gets an actuator handle successfully.

**Comment by destogl** [comment]
Hi @bmagyar,

I see the reasons you need this for further development on Transmissions, nevertheless, IMO, this functionality would be in the future part of the Actuator and Sensor-Components. So that component would create transmission class for themselves and use it.

Can this PR see as a temporerly solution? If so, I propose to rename `ActuatorHandle` to something elese so it does not interfere/confuse the implementation of components.

File: hardware_interface/test/test_register_actuators.cpp (hunk 1 above)
Comment (bmagyar) [reply hardware_interface/test/test_register_actuators.cpp:162]: added

**Comment by bmagyar** [comment]
I wouldn't necessarily call it a temporary solution but rather a minimal first *enabler* version which could be overridden, redesigned to a full extent. How do you propose to rename it? I'd be happy for you to revamp these handles in the future but I wouldn't want to add them as something very different from actuators as that's exactly what they are... Perhaps something to pinpoint that they just handle single actuator interface values? `ActuatorValueHandle`?

**Comment by Karsten1987** [comment]
I would hold this PR back until #101 is merged. I don't see the point of introducing a temporary interface which then has to be deleted afterwards.

File: hardware_interface/test/test_register_actuators.cpp (hunk 1 above)
Comment (jordan-palacios) [reply hardware_interface/test/test_register_actuators.cpp:162]: Thanks.

**Comment by bmagyar** [comment]
When is #101 ready for a merge?
Can you outline how you'd implement the same thing with #101 ? Are you going to incorporate the logic from here in there?

**Comment by jordan-palacios** [comment]
> I would hold this PR back until #101 is merged. I don't see the point of introducing a temporary interface which then has to be deleted afterwards.

I disagree. This PR is ready now and will allow us to start working on all the Trasnmission related code: `TrasnmissionInterface`, `TrasnmissionLoader`, `SimpleTransmission`, etc. 

I love the work being poured in #101 (I'm myself trying to catch up on it to weigh in) but as of now is probably weeks away from being ready to be part of the master branch.

**Comment by Karsten1987** [comment]
> This PR is ready now and will allow us to start working on all the Trasnmission related code: TrasnmissionInterface, TrasnmissionLoader, SimpleTransmission, etc.

I guess I am missing something here, but I don't see how this PR is related to transmissions? The interfaces added here are introducing design which is all fundamentally revoked by #101. The resource management is planned to be different, the memory allocation as well as composing multiple actuators into a "system" of actuators.

> Are you going to incorporate the logic from here in there?

As mentioned above, I don't see which logic you mean. The resource management? That'll be addressed after #101 is merged - IMO. But asking the other way around, if we merge this before #101 are you incorporating this logic here into the new setup?

But that's just my take on this. Happy to be overruled ;-)

**Comment by bmagyar** [comment]
Following our meeting yesterday I am going ahead with merging this. 

The CI has multiple issues at this point: failing to find control_msgs as it is looking up binary packages in eloquent and ignoring the .repos file (which is either in the wrong spot or is being removed (not sure about one bit on the CI messages)). I'm giving it another go, forcing foxy to be picked up for binaries but perhaps rolling would be more appropriate (can discuss later)

**Comment by bmagyar** [comment]
Apart from the codecov glitches, this PR is ready to go, merging...

**Pull request merged**